File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/dagger/modules/PresenterModule.java
package com.test.amaro.amarotest.dagger.modules;

import android.content.SharedPreferences;
import com.test.amaro.amarotest.dagger.scopes.PerActivity;
import com.test.amaro.amarotest.data.usecases.ProductUseCase;
import com.test.amaro.amarotest.presentation.detail.DetailPresenter;
import com.test.amaro.amarotest.presentation.main.MainPresenter;
import com.test.amaro.amarotest.presentation.main.filter.FilterPresenter;
import dagger.Module;
import dagger.Provides;

@Module
public final class PresenterModule {

    @Provides
    @PerActivity
    public final MainPresenter providesMainPresenter(ProductUseCase productUseCase) {
        return new MainPresenter(productUseCase);
    }

    @Provides
    @PerActivity
    public final DetailPresenter provideDetailPresenter() {
        return new DetailPresenter();
    }

    @Provides
    @PerActivity
    public final FilterPresenter provideFilterPresenter(SharedPreferences sharedPreferences) {
        return new FilterPresenter(sharedPreferences);
    }
}

File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/dagger/modules/PresenterModule.kt
package com.test.amaro.amarotest.dagger.modules

import com.test.amaro.amarotest.dagger.scopes.PerActivity
import com.test.amaro.amarotest.data.repository.ProductRepository
import com.test.amaro.amarotest.main.MainPresenter
import dagger.Module
import dagger.Provides

@Module
class PresenterModule {

    @Provides
    @PerActivity
    fun providesMainPresenter(productRepository: ProductRepository): MainPresenter {
        return MainPresenter(productRepository)
    }
}